disco_aws_automation/disco_ssm.py

```diff
@@ -0,0 +1,175 @@
+    """
+
+    def __init__(self, environment_name=None, config_aws=None):
+        self.config_aws = config_aws or read_config()
+
+        if environment_name:
+            self.environment_name = environment_name.lower()
+        else:
+            self.environment_name = self.config_aws.get("disco_aws", "default_environment")
+
+        self._conn = None  # Lazily initialized
+
+    @property
+    def conn(self):
+        """The boto3 ssm connection object"""
+        if not self._conn:
+            self._conn = boto3.client('ssm')
+        return self._conn
+
+    def get_all_documents(self):
+        """ Returns a list of existing SSM documents."""
+        next_token = ''
+        documents = []
+        while True:
+            if next_token:
+                response = throttled_call(self.conn.list_documents, NextToken=next_token)
+            else:
+                response = throttled_call(self.conn.list_documents)
+
+            documents.extend(response.get("DocumentIdentifiers"))
+            next_token = response.get("NextToken")
+
+            if not next_token:
+                break
+
+        result = [doc for doc in documents if doc["Name"].startswith(DOCUMENT_PREFIX)]
+        return result
+
+    def get_document_content(self, doc_name):
+        """ Returns the content of the document."""
+        if not doc_name.startswith(DOCUMENT_PREFIX):
+            raise Exception("Document name ({0}) doesn't start with \"{1}\".".
+                            format(doc_name, DOCUMENT_PREFIX))
+
+        try:
+            response = throttled_call(self.conn.get_document, Name=doc_name)
+        except ClientError:
+            logger.info("Document name (%s) is not found.", doc_name)
+            return None
+
+        return response.get("Content")
+
+    def update(self, wait=False, dry_run=False):
+        """ Updates SSM documents from configuration """
+        desired_docs = set(self._list_docs_in_config())
+        existing_docs = set([doc["Name"] for doc in self.get_all_documents()])
+
+        docs_to_create = desired_docs - existing_docs
+        docs_to_delete = existing_docs - desired_docs
+        docs_to_update = self._check_for_update(desired_docs & existing_docs)
+        unchanged_docs = existing_docs - docs_to_update - docs_to_delete
+
+        logger.info("New documents to be added: %s", docs_to_create)
+        logger.info("Documents to be deleted: %s", docs_to_delete)
+        logger.info("Existing documents to be updated: %s", docs_to_update)
+        logger.info("Unchanged documents: %s", unchanged_docs)
+
+        if not dry_run:
+            # Include docs_to_update in docs_to_delete so that they can be recreated later
+            docs_to_delete |= docs_to_update
+            self._delete_docs(docs_to_delete)
+            if wait:
+                self._wait_for_docs_deleted(docs_to_delete)
+
+            docs_to_create |= docs_to_update
+            self._create_docs(docs_to_create)
+            if wait:
```

Comment:
I was just thinking letting the user of the tool to decide whether to wait or not. Maybe we change to the default behavior to wait?